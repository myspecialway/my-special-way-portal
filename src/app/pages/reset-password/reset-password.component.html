<div class="login-container">
  <form name="form" (ngSubmit)="f.form.valid && resetPassword()" #f="ngForm" novalidate>
    <div class="login-card">
      <div class="msw-login-logo-ph"></div>
      <div class="msw-login-title">בדרכי שלי</div>
      <div class="msw-reset-title">שכחת סיסמא ?</div>
      <div class="msw-reset-subtitle">אין בעייה, הזן את כתובת המייל ואנו נשלח לך סיסמא חדשה</div>
      <div class="card-form" [ngClass]="{ 'has-error': f.submitted && !email.valid }">
        <div class="form-row">
          <mat-form-field
            [hideRequiredMarker]="formFieldOptions.value.hideRequired"
            [floatLabel]="formFieldOptions.value.floatLabel">
            <input type="email" matInput placeholder="אימייל" name="email" [(ngModel)]="model.email" #email="ngModel" required/>
          </mat-form-field>
          <div *ngIf="f.submitted && !email.valid" class="help-block" data-test-id="empty-email-err">נא הכנס דוא״ל</div>
        </div>
        <a class="nav-link" routerLink="/login">נזכרתי בסיסמא</a>
      </div>
      <div class="card-footer">
        <button mat-button>שחזר סיסמא</button>
        <!--<button mat-button (click)="loginBtn()">הרשמה</button>-->
      </div>
    </div>
  </form>
</div>

