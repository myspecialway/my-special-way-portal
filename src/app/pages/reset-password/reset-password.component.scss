:host {
  height: 100%;
  background-color: #fff;
  display: block;
  position: fixed;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  z-index: 9999;
}
.login-container {
  background-color: #fff;
  height: 100%;

  @-webkit-keyframes card {
    from {
      top: -40px;
    }
    to {
      top: 0;
    }
  }

  @keyframes card {
    from {
      top: -40px;
    }
    to {
      top: 0;
    }
  }
  .login-card {
    -moz-animation-duration: 600ms;
    -moz-animation-fill-mode: forwards;
    -moz-animation-name: card;
    -o-animation-duration: 600ms;
    -o-animation-fill-mode: forwards;
    -o-animation-name: card;
    -webkit-animation-duration: 600ms;
    -webkit-animation-fill-mode: forwards;
    -webkit-animation-name: card;
    align-items: center;
    animation-duration: 600ms;
    animation-fill-mode: forwards;
    animation-name: card;
    display: flex;
    flex-direction: column;
    padding: 20px;
    position: absolute;
    width: 520px;
    top: 50%;
    left: 50%;
    transform: translate(-50%);
    margin-top: 100px;

    .msw-login-logo-ph {
      margin-bottom: 50px;
    }

    .msw-login-title {
      margin-top: 20px;
      height: 24px;
      font-size: 20px;
      font-weight: 500;
      letter-spacing: 0.4px;
      color: #6a00cf;
    }

    .msw-login-subtitle {
      margin-top: 12px;
      font-size: 21px;
      letter-spacing: 0.4px;
      color: #222222;
    }

    .msw-reset-subtitle {
      font-size: 14px;
    }

    .msw-reset-error {
      margin-top: 12px;
      font-size: 14px;
      letter-spacing: 0.4px;
      color: #b71c1c;
    }

    .form-row,
    .card-form,
    .mat-form-field {
      width: 100%;
    }

    .card-form {
      padding: 15px 80px;

      .form-row {
        position: relative;
        display: flex;
        align-items: center;
        margin-top: 13px;
      }
    }
    .card-footer {
      align-items: center;
      border: 0;
      display: flex;
      justify-content: space-between;
      margin: 0;
      padding: 0 80px;
      width: 100%;
      button {
        background-color: #6a00cf;
        border-radius: 3px;
        color: #fff;
        height: 36px;
        width: 300px;
      }
    }

    .msw-login-checkbox {
      .mat-checkbox-checked {
        /deep/ .mat-checkbox-background {
          background-color: #6a00cf;
        }
      }
    }

    .msw-reset-title {
      font-size: 21px;
    }
  }

  .mat-form-field-label-wrapper {
    border: 1px solid #f0f;
  }
}
