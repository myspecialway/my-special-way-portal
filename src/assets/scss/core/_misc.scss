body {
  background-color: #fafafa;
  color: $black-color;
  font-weight: 300;
}

legend {
  border-bottom: 0;
}

.serif-font {
  font-family: $font-family-serif;
}

// Prevent highlight on mobile
* {
  -webkit-tap-highlight-color: rgba(255, 255, 255, 0);
  -webkit-tap-highlight-color: transparent;
  &:focus {
    outline: 0;
  }
}

a {
  color: $link-color;
  &:hover,
  &:focus {
    color: darken($link-color, 5%);
    text-decoration: none;
  }

  &.text-info {
    &:hover,
    &:focus {
      color: darken($brand-info, 5%);
    }
  }

  & .material-icons {
    vertical-align: middle;
  }
}

.form-check,
label {
  font-size: 14px;
  line-height: 1.42857;
  color: $checkboxes-text-color;
  font-weight: 400;
}

/*           Animations              */
.animation-transition-general {
  @include transition-all($general-transition-time, $transition-linear);
}

.animation-transition-slow {
  @include transition-all($slow-transition-time, $transition-linear);
}

.animation-transition-fast {
  @include transition-all($fast-transition-time, $transition-ease);
}

.caret,
.sidebar a {
  @include transition-all($fast-transition-time, $transition-ease-in);
}
#map {
  position: relative;
  z-index: 2;
  width: 100%;
  height: calc(100vh - 70px);
  margin-top: 70px;
}

@font-face {
  font-family: 'pass';
  font-style: normal;
  font-weight: 400;
  src: url(data:application/font-woff;charset=utf-8;base64,d09GRgABAAAAAATsAA8AAAAAB2QAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABWAAAABwAAAAcg9+z70dERUYAAAF0AAAAHAAAAB4AJwANT1MvMgAAAZAAAAA/AAAAYH7AkBhjbWFwAAAB0AAAAFkAAAFqZowMx2N2dCAAAAIsAAAABAAAAAQAIgKIZ2FzcAAAAjAAAAAIAAAACAAAABBnbHlmAAACOAAAALkAAAE0MwNYJ2hlYWQAAAL0AAAAMAAAADYPA2KgaGhlYQAAAyQAAAAeAAAAJAU+ATJobXR4AAADRAAAABwAAAAcCPoA6mxvY2EAAANgAAAAEAAAABAA5gFMbWF4cAAAA3AAAAAaAAAAIAAKAE9uYW1lAAADjAAAARYAAAIgB4hZ03Bvc3QAAASkAAAAPgAAAE5Ojr8ld2ViZgAABOQAAAAGAAAABuK7WtIAAAABAAAAANXulPUAAAAA1viLwQAAAADW+JM4eNpjYGRgYOABYjEgZmJgBEI2IGYB8xgAA+AANXjaY2BifMg4gYGVgYVBAwOeYEAFjMgcp8yiFAYHBl7VP8wx/94wpDDHMIoo2DP8B8kx2TLHACkFBkYA8/IL3QB42mNgYGBmgGAZBkYGEEgB8hjBfBYGDyDNx8DBwMTABmTxMigoKKmeV/3z/z9YJTKf8f/X/4/vP7pldosLag4SYATqhgkyMgEJJnQFECcMOGChndEAfOwRuAAAAAAiAogAAQAB//8AD3jaY2BiUGJgYDRiWsXAzMDOoLeRkUHfZhM7C8Nbo41srHdsNjEzAZkMG5lBwqwg4U3sbIx/bDYxgsSNBRUF1Y0FlZUYBd6dOcO06m+YElMa0DiGJIZUxjuM9xjkGRhU2djZlJXU1UDQ1MTcDASNjcTFQFBUBGjYEkkVMJCU4gcCKRTeHCk+fn4+KSllsJiUJEhMUgrMUQbZk8bgz/iA8SRR9qzAY087FjEYD2QPDDAzMFgyAwC39TCRAAAAeNpjYGRgYADid/fqneL5bb4yyLMwgMC1H90HIfRkCxDN+IBpFZDiYGAC8QBbSwuceNpjYGRgYI7594aBgcmOAQgYHzAwMqACdgBbWQN0AAABdgAiAAAAAAAAAAABFAAAAj4AYgI+AGYB9AAAAAAAKgAqACoAKgBeAJIAmnjaY2BkYGBgZ1BgYGIAAUYGBNADEQAFQQBaAAB42o2PwUrDQBCGvzVV9GAQDx485exBY1CU3PQgVgIFI9prlVqDwcZNC/oSPoKP4HNUfQLfxYN/NytCe5GwO9/88+/MBAh5I8C0VoAtnYYNa8oaXpAn9RxIP/XcIqLreZENnjwvyfPieVVdXj2H7DHxPJH/2/M7sVn3/MGyOfb8SWjOGv4K2DRdctpkmtqhos+D6ISh4kiUUXDj1Fr3Bc/Oc0vPqec6A8aUyu1cdTaPZvyXyqz6Fm5axC7bxHOv/r/dnbSRXCk7+mpVrOqVtFqdp3NKxaHUgeod9cm40rtrzfrt2OyQa8fppCO9tk7d1x0rpiQcuDuRkjjtkHt16ctbuf/radZY52/PnEcphXpZOcofiEZNcQAAeNpjYGIAg///GBgZsAF2BgZGJkZmBmaGdkYWRla29JzKggxD9tK8TAMDAxc2D0MLU2NjENfI1M0ZACUXCrsAAAABWtLiugAA)
    format('woff');
}

input[type='password'] {
  font-family: 'pass', 'Roboto', Helvetica, Arial, sans-serif;
  font-size: 18px;
  &::-webkit-input-placeholder {
    transform: scale(0.77);
    transform-origin: 0 50%;
  }
  &::-moz-placeholder {
    font-size: 14px;
    opacity: 1;
  }
  &:-ms-input-placeholder {
    font-size: 14px;
    font-family: 'Roboto', Helvetica, Arial, sans-serif;
  }
}
