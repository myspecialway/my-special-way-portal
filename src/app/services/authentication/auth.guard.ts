import gql from 'graphql-tag';
import { Injectable } from '@angular/core';
import { ActivatedRouteSnapshot, CanActivate, Router, RouterStateSnapshot, UrlSegment } from '@angular/router';
import { AuthenticationService } from './authentication.service';
import { UserType } from '../../models/user.model';
import { Apollo } from 'apollo-angular';
import { UserProfileStateModel } from '../../apollo/state/resolvers/state.resolver';

@Injectable()
export class AuthGuard implements CanActivate {
  constructor(
    private router: Router,
    private authService: AuthenticationService,
    private apollo: Apollo,
  ) { }

  async canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<boolean> {
    await this.authService.checkRestoreAuthData();
    const response = await this.apollo.query<{ userProfile: UserProfileStateModel }>({
      query: gql`
      query {
        userProfile @client{
          role
          token
        }
      }
    `}).toPromise();

    if (
      response.data.userProfile &&
      response.data.userProfile.token &&
      !this.authService.isTokenExpired(response.data.userProfile.token)
      && this.isAuthorized(route.url, response.data.userProfile.role)
    ) {
      return true;
    }

    this.router.navigate(['/login'], { queryParams: { returnUrl: state.url } });
    return false;
  }

  isAuthorized(route: UrlSegment[], userRole: UserType) {
    // console.log(`route ${route} | User type ${userRole}`);
    // if (route[0].path.toString() === 'user' && userRole.toString() === 'TEACHER') {
    //   return false;
    // }
    return true; // TODO: authorization logic
  }
}
