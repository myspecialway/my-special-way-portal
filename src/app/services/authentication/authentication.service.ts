import 'rxjs/add/operator/map';
import 'rxjs/add/operator/catch';
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { LoginResponse } from '../../models/login-response.model';
import { environment } from '../../../environments/environment';
import { JwtHelperService } from '@auth0/angular-jwt';
import { JWTTokenPayloadResponse } from '../../models/jwt-token-resonse.model';
import { Apollo } from 'apollo-angular';
import { UPDATE_USER_PROFILE } from '../../apollo/state/mutations/update-user-profile.mutation';
import { UserProfileStateModel } from '../../apollo/state/resolvers/state.resolver';
import { UserUniqueValidationResponse } from '../../models/user-unique-validation-response.model';
import { of } from 'rxjs';
import { catchError, first, map } from 'rxjs/operators';

@Injectable()
export class AuthenticationService {
  private jwtHelper = new JwtHelperService();
  private isStateRestored = false;

  constructor(private http: HttpClient, private apollo: Apollo) {}

  private getTokenFromLocalStore() {
    return localStorage.getItem('token') ? localStorage.getItem('token') : sessionStorage.getItem('token');
  }

  async checkRestoreAuthData() {
    if (this.isStateRestored) {
      return;
    }

    this.isStateRestored = true;
    const token = this.getTokenFromLocalStore();

    if (!token) {
      return;
    }

    const userProfile = this.getProfileFromToken(token);
    await this.pushUserProfileToState(userProfile);
  }

  async login(username: string, password: string, isRememberMeActive: boolean): Promise<boolean> {
    try {
      const tokenResponse = await this.http
        .post<LoginResponse>(environment.hotConfig.MSW_HOT_LOGIN_ENDPOINT, { username, password })
        .toPromise();

      if (!tokenResponse) {
        return false;
      }

      if (isRememberMeActive) {
        localStorage.setItem('token', tokenResponse.accessToken);
      } else {
        sessionStorage.setItem('token', tokenResponse.accessToken);
      }

      const userProfile = this.getProfileFromToken(tokenResponse.accessToken);
      await this.pushUserProfileToState(userProfile);

      return true;
    } catch (error) {
      const typedError = error as HttpErrorResponse;

      if (typedError.status !== 401) {
        throw error;
      }

      return false;
    }
  }

  private getProfileFromToken(token: string): UserProfileStateModel {
    const jwrParsedToken = this.parseToken(token);
    const userProfile = new UserProfileStateModel(jwrParsedToken);
    userProfile.token = token;

    return userProfile;
  }

  private parseToken(token: string): JWTTokenPayloadResponse {
    const decodedToken = this.jwtHelper.decodeToken(token) as JWTTokenPayloadResponse;
    return decodedToken;
  }

  private async pushUserProfileToState(userProfile: UserProfileStateModel) {
    return this.apollo
      .mutate({
        mutation: UPDATE_USER_PROFILE,
        variables: {
          userProfile,
        },
      })
      .toPromise();
  }

  isTokenExpired(token: string): boolean {
    return this.jwtHelper.isTokenExpired(token);
  }

  logout() {
    localStorage.removeItem('token');
    sessionStorage.removeItem('token');
  }

  checkUsernameUnique(username: string, id: string) {
    if (!username) {
      return of(true);
    }
    const userID = id === '' ? undefined : id;
    try {
      return this.http
        .post<UserUniqueValidationResponse>(environment.hotConfig.MSW_HOT_USER_UNIQUE_VALIDATION_ENDPOINT, {
          username,
          id: userID,
        })
        .pipe(
          first(),
          map((res) => res.isUnique),
          catchError(() => of(true)),
        );
    } catch (error) {
      return of(true);
    }
  }
}
