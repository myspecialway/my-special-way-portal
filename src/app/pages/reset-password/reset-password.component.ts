import { Component, OnInit } from '@angular/core';
import { FormBuilder, FormGroup } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { AuthenticationService } from '../../services/authentication/authentication.service';

@Component({
  selector: 'app-reset-password',
  templateUrl: './reset-password.component.html',
  styleUrls: ['./reset-password.component.scss'],
})
export class ResetPasswordComponent implements OnInit {
  model: any = {};
  returnUrl: string;
  formFieldOptions: FormGroup;
  sentResponse = false;
  errorMessage = '';
  constructor(
    private route: ActivatedRoute,
    private router: Router,
    private authenticationService: AuthenticationService,
    formBuilder: FormBuilder,
  ) {
    this.formFieldOptions = formBuilder.group({ hideRequired: true });
  }

  ngOnInit(): void {
    this.authenticationService.logout();
    this.returnUrl = this.route.snapshot.queryParams['returnUrl'] || '/';
  }

  async resetPassword() {
    try {
      const sentResponse = await this.authenticationService.resetPassword(this.model.email);

      if (!sentResponse) {
        console.warn('reset-password.component::reset-password:: reset-password error');
        return;
      }

      this.router.navigate(['/sent-successfully']);
    } catch (err) {
      console.error(`reset-password.component::reset-password:: error in authentication ${err}`);
      this.errorMessage = 'לא הצלחנו לשלוח אימייל לכתובת: ' + this.model.email + '. אנא נסה שוב.';
    }
  }
}
