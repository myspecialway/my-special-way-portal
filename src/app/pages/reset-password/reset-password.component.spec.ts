import { async, TestBed } from '@angular/core/testing';
import { ActivatedRoute, Router, RouterModule } from '@angular/router';
import { ResetPasswordComponent } from './reset-password.component';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { AuthenticationService } from '../../services/authentication/authentication.service';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';

describe('ResetPasswordComponent', () => {
  let authenticationServiceMock: Partial<AuthenticationService>;
  let routerModuleMock: Partial<RouterModule>;

  beforeEach(async(() => {
    authenticationServiceMock = {
      logout: jest.fn(),
      resetPassword: jest.fn(),
    };

    routerModuleMock = {
      forRoot: jest.fn(),
      forChild: jest.fn(),
      navigate: jest.fn(),
    };

    TestBed.configureTestingModule({
      imports: [FormsModule, ReactiveFormsModule],
      declarations: [ResetPasswordComponent],
      providers: [
        {
          provide: ActivatedRoute,
          useValue: {
            snapshot: {
              queryParams: {
                returnUrl: 'blabla',
              },
            },
          },
        },
        { provide: AuthenticationService, useValue: authenticationServiceMock },
        { provide: Router, useValue: routerModuleMock },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });
  }));

  it('should render the component as described in snapshot', () => {
    const fixture = TestBed.createComponent(ResetPasswordComponent);
    expect(fixture).toMatchSnapshot();
  });

  it('should call resetPassword function on resetPassword', async () => {
    const fixture = TestBed.createComponent(ResetPasswordComponent);
    fixture.componentInstance.resetPassword();
    expect(authenticationServiceMock.resetPassword).toHaveBeenCalled();
  });

  it('should call logout function on init', async () => {
    const fixture = TestBed.createComponent(ResetPasswordComponent);
    fixture.componentInstance.ngOnInit();
    expect(authenticationServiceMock.logout).toHaveBeenCalled();
  });
});
