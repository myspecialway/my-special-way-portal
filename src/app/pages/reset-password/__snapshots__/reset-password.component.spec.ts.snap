// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPasswordComponent should render the component as described in snapshot 1`] = `
<app-reset-password
  authenticationService={[Function Object]}
  formFieldOptions={[Function FormGroup]}
  model={[Function Object]}
  route={[Function Object]}
  router={[Function Object]}
  sentResponse="false"
>
  <div
    class="login-container"
  >
    <form
      name="form"
      novalidate=""
    >
      <div
        class="login-card"
      >
        <div
          class="msw-login-logo-ph"
        >
          <img
            alt=""
            src="/assets/img/myway.png"
          />
        </div>
        <div
          class="msw-reset-title"
        >
          שכחת סיסמא ?
        </div>
        <div
          class="msw-reset-subtitle"
        >
           אין בעייה, הזן את כתובת המייל ואנו נשלח לך סיסמא חדשה 
        </div>
        <div
          class="card-form"
        >
          <div
            class="form-row"
          >
            <mat-form-field>
              <input
                matinput=""
                name="email"
                placeholder="אימייל"
                required=""
                type="email"
              />
            </mat-form-field>
            
          </div>
          <a
            class="back-to-login"
            routerlink="/login"
          >
            נזכרתי בסיסמא
          </a>
        </div>
        <div
          class="card-footer"
        >
          <button
            mat-button=""
          >
            שחזר סיסמא
          </button>
        </div>
      </div>
    </form>
    <div
      class="msw-login-footer"
    >
      <img
        class="img"
        src="/assets/img/tinti.png"
      />
      <div
        class="gray-bar"
      />
    </div>
  </div>
</app-reset-password>
`;
